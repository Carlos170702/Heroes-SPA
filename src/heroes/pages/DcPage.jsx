import { HeroList } from "../components"

export const DcPage = () => {
  const publisher = 'DC Comics';
  return (
    <>
      <h1>DC Page</h1>
      <hr />

      <ul>
        {
          <HeroList publisher={publisher} />
        }
      </ul>

    </>
  )
}
