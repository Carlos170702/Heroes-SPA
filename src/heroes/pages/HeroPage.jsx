import { useMemo } from 'react';
import { Navigate, useNavigate, useParams } from 'react-router-dom';
import { getHeroById } from '../helpers/getHeroById';

export const HeroPage = () => {
  const { id } = useParams();
  const navigate = useNavigate();

  const heroe =useMemo( () => getHeroById(id),[id]) ;

  const heroeImage = `/assets/heroes/${id}.jpg`

  const onBackPage =() => {
    navigate( -1 );
  }

  if (!heroe) {
    return <Navigate to="/marvel" />
  }

  return (
    <>
      <div className='fullHero animate__animated animate__bounceIn '>
        <div className="heroData">

          <div className="imgHeroes animate__rotateOut">
                <img src={heroeImage} alt={heroe.superhero} className="imgHeroe" />
          </div>

          <div className='infHeroes'>
            <h3 className="card-tittle" >{ heroe.superhero}</h3>
            <ul className='list' >
              <li className='list-item'><b>Alter Ego: </b> { heroe.alter_ego } </li>
              <li className='list-item'><b>Publisher: </b> { heroe.publisher } </li>
              <li className='list-item'><b>First appearance: </b> { heroe.first_appearance } </li>
            </ul>

            <h5 className='mt-3'>Characters</h5>
            <p>{ heroe.characters }</p>


            <button 
              className='btn btn-outline-primary mt-10'
              onClick={ onBackPage}
            >
              Return 
            </button>
          </div>
          
        </div>
      </div>
    </>
  )
}
