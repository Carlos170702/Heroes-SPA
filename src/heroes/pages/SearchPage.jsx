import { useForm } from "../../hooks/useForm";
import querystring from 'query-string'; //se installa query-string
import { useNavigate, useLocation } from 'react-router-dom';
import { getHeroesByName } from "../helpers";
import { HeroCard } from "../components/HeroCard";


export const SearchPage = () => {

  const navigate = useNavigate();
  //sirve para traer la localisacion donde estoy
  const location = useLocation();

  //se usa query-string se parsea el arreglo de localisacion
  const { q = '' } = querystring.parse(location.search);
  const heroes = getHeroesByName(q)

  const { formState: { searchText }, onInputChange } = useForm({
    searchText: q
  });

  const onSubmit = (e) => {
    e.preventDefault();
    //esto navega ala misma pagina pero con un queryparameters 
    navigate(`?q=${searchText}`);
  }

  return (
    <>
      <h1>SearchPage</h1>
      <hr />

      <div className="search">

        <div className="search-heroes">
          <div className="search-heroe">
            <h4>Searching</h4>
            <form
            aria-label="form"
              className="mt-2"
              onSubmit={onSubmit}
            >
              <input
                type="text"
                placeholder="Search a hero "
                className="form-control mt-2"
                name="searchText"
                autoComplete="off"
                onChange={onInputChange}
                value={searchText}
              />
              <button
                className="btn btn-outline-primary mt-2"
              >
                Search
              </button>
            </form>
          </div>

          <div className="results-heroe">
            <h4>Results</h4>

            {/* {
              (q === '')
                ? <div className="alert alert-primary mt'2">Search Hero</div>
                : (heroes.length === 0) && <div className="alert alert-danger mt-2">no hero with <b>{q}</b></div>
            } */}


            <div
              aria-label="divNone"
              className="alert alert-primary mt-2 animate__animated animate__fadeIn"
              style={{ display: q === '' ? '' : 'none' }}
            >
              Search Hero
            </div>

            <div
              aria-label="divNoHero"
              className="alert alert-danger mt-2 animate__animated animate__fadeIn"
              style={{ display: heroes.length === 0 && q.length > 0 ? '' : 'none' }}
            >
              no hero with <b>{q}</b>
            </div>

            {
              heroes.map(heroe => (
                < HeroCard key={heroe.id} {...heroe} />
              ))
            }
          </div>
        </div>

      </div>
    </>
  )
}
