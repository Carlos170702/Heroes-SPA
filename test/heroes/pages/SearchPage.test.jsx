import { fireEvent, render, screen } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    //que extraiga todo lo que tiene el react-router-dom
    ...jest.requireActual('react-router-dom'),
    //y solo se sobrescribe esta funcion
    useNavigate: () => mockedUseNavigate
}))

describe('pruebas en <SearchPage />', () => {
    beforeEach( () => jest.clearAllMocks() )
    test('debe de mostrarse correctamente con valores por defecto', () => {

        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();

    })


    test('debe mostrar a Batman y el input con el valor de queryString ', () => {
        render(
            <MemoryRouter initialEntries={['search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        )

        const input = screen.getByRole('textbox');
        expect(input.value).toBe('batman')

        const img = screen.getByRole('img');
        expect(img.src).toContain("/assets/heroes/dc-batman.jpg")

        const divNone = screen.getByLabelText('divNone')
        const { style: { display } } = divNone
        expect(display).toBe('none');
    })

    test('debe de mostrar un error si no encuentra el heroe (batman123)', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>
        )

        const divNone = screen.getByLabelText('divNoHero')
        expect(divNone.style.display).toBe('');
    })

    test('debe de llamar al navigate a la pantalla nueva', () => {
        const inputValue = 'batman';
        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        )

        // screen.debug();

        const input = screen.getByRole('textbox');
        fireEvent.change(input, { target: { name: 'searchText', value: inputValue} })

        const form = screen.getByRole('form')
        fireEvent.submit(form)

        expect(mockedUseNavigate).toHaveBeenCalled();
        expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${ inputValue}`);

    })
})
