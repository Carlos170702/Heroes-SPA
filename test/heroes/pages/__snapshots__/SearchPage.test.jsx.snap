// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <SearchPage /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    SearchPage
  </h1>
  <hr />
  <div
    class="search"
  >
    <div
      class="search-heroes"
    >
      <div
        class="search-heroe"
      >
        <h4>
          Searching
        </h4>
        <form
          aria-label="form"
          class="mt-2"
        >
          <input
            autocomplete="off"
            class="form-control mt-2"
            name="searchText"
            placeholder="Search a hero "
            type="text"
            value=""
          />
          <button
            class="btn btn-outline-primary mt-2"
          >
            Search
          </button>
        </form>
      </div>
      <div
        class="results-heroe"
      >
        <h4>
          Results
        </h4>
        <div
          aria-label="divNone"
          class="alert alert-primary mt-2 animate__animated animate__fadeIn"
        >
          Search Hero
        </div>
        <div
          aria-label="divNoHero"
          class="alert alert-danger mt-2 animate__animated animate__fadeIn"
          style="display: none;"
        >
          no hero with 
          <b />
        </div>
      </div>
    </div>
  </div>
</div>
`;
